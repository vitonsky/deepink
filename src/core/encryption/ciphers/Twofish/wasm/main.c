#include	"aes.h"
#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<time.h>
#include	<ctype.h>

int main() {
    cipherInstance cipher;
    BYTE iv[16] = {0};
    cipherInit(&cipher, MODE_ECB, iv);
    // call encrypt/decrypt functions for testing
    return 0;
}