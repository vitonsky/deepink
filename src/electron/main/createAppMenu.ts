/* eslint-disable spellcheck/spell-checker */
import { app, Menu, MenuItem, MenuItemConstructorOptions } from 'electron';

import { openAboutWindow } from '../../windows/about';

export function createAppMenu() {
	type MenuObject = MenuItemConstructorOptions | MenuItem;
	const application: MenuObject = {
		label: 'Application',
		role: 'about',
		submenu: [
			{
				label: 'About Application',
				click(_item, window) {
					openAboutWindow(window);
				},
			},
			{
				type: 'separator',
			},
			{
				label: 'Quit',
				accelerator: 'Command+Q',
				click: () => {
					app.quit();
				},
			},
		],
	};

	const edit: MenuObject = {
		label: 'Edit',
		submenu: [
			{
				label: 'Undo',
				accelerator: 'CmdOrCtrl+Z',
				role: 'undo',
			},
			{
				label: 'Redo',
				accelerator: 'Shift+CmdOrCtrl+Z',
				role: 'redo',
			},
			{
				type: 'separator',
			},
			{
				label: 'Cut',
				accelerator: 'CmdOrCtrl+X',
				role: 'cut',
			},
			{
				label: 'Copy',
				accelerator: 'CmdOrCtrl+C',
				role: 'copy',
			},
			{
				label: 'Paste',
				accelerator: 'CmdOrCtrl+V',
				role: 'paste',
			},
			{
				label: 'Select All',
				accelerator: 'CmdOrCtrl+A',
				role: 'selectAll',
			},
		],
	};

	return Menu.buildFromTemplate([
		application,
		edit,
		{
			label: 'Help',
			role: 'help',
			submenu: [
				{
					label: 'Toggle dev tools',
					role: 'toggleDevTools',
					accelerator: 'CmdOrCtrl+Alt+Shift+I',
				},
			],
		},
	]);
}
