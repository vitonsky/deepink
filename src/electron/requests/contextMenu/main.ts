import { BrowserWindow, Menu } from 'electron';

import { ipcMainHandler } from '../../utils/ipc/ipcMainHandler';

import { contextMenuChannel } from '.';

export const enableContextMenu = () =>
	contextMenuChannel.server(ipcMainHandler, {
		async open({ req: [props], ctx: event }) {
			const targetWindow = BrowserWindow.fromWebContents(event.sender);
			if (!targetWindow) return;

			const { menu: menuTemplate, x = 0, y = 0 } = props ?? {};

			// Handle click on menu
			let onClick: (itemId: string | null) => void;
			const onClickPromise = new Promise<string | null>((res) => {
				onClick = res;
			});

			const prepareMenu = (menu: Electron.MenuItemConstructorOptions[]) => {
				return menu.map(
					({ id, click, ...props }): Electron.MenuItemConstructorOptions => {
						// Decorate click handler, to resolve promise with menu id by click on menu
						const clickHandler =
							id === undefined
								? click
								: (
										menuItem: Electron.MenuItem,
										browserWindow: Electron.BaseWindow | undefined,
										event: Electron.KeyboardEvent,
								  ) => {
										if (click) click(menuItem, browserWindow, event);
										onClick(id);
								  };

						return { id, click: clickHandler, ...props };
					},
				);
			};

			const menu = Menu.buildFromTemplate(prepareMenu(menuTemplate));

			menu.popup({ window: targetWindow, x, y });
			menu.once('menu-will-close', () => {
				// Delay on next tick, to allow to handle click event
				setTimeout(() => {
					onClick(null);
				}, 0);
			});

			return onClickPromise;
		},
	});
