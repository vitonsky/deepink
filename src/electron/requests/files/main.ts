import { BrowserWindow, dialog } from 'electron';
import { globSync } from 'fast-glob';
import { lstatSync } from 'fs';

import * as filesUtils from '../../utils/files';
import { ipcMainHandler } from '../../utils/ipc/ipcMainHandler';

import { readFile, realpath } from 'fs/promises';
import { filesChannel } from '.';

export const serveFiles = () =>
	filesChannel.server(ipcMainHandler, {
		async getUserDataPath({ req: [path] }) {
			return filesUtils.getUserDataPath(path);
		},

		async getResourcesPath({ req: [path] }) {
			return filesUtils.getResourcesPath(path);
		},

		async selectDirectory({ ctx: evt }) {
			const window = BrowserWindow.getAllWindows().find(
				(win) => win.webContents.id === evt.sender.id,
			);
			if (!window) return null;

			const { filePaths } = await dialog.showOpenDialog(window, {
				properties: ['openDirectory'],
			});

			return filePaths;
		},

		async importNotes({ ctx: evt }) {
			const window = BrowserWindow.getAllWindows().find(
				(win) => win.webContents.id === evt.sender.id,
			);
			if (!window) return;

			console.warn('Start export');
			const { filePaths } = await dialog.showOpenDialog(window, {
				properties: ['openDirectory'],
			});
			if (filePaths.length === 0) return;

			const directoryToScan = filePaths[0];
			const files = globSync(`${directoryToScan}/{*,**/*}`);

			const rootDir = await realpath(directoryToScan);
			const filesMap: Record<string, ArrayBuffer> = {};
			await Promise.all(
				files.map(async (file) => {
					const absoluteFilename = await realpath(file);
					if (!absoluteFilename.startsWith(rootDir)) return;
					if (!lstatSync(file).isFile()) return;

					const rootPathLengthWithSlash = rootDir.length + 1;
					const filename = absoluteFilename.slice(rootPathLengthWithSlash);
					const buffer = await readFile(file);
					filesMap[filename] = new Uint8Array(buffer.buffer)
						.buffer as ArrayBuffer;
				}),
			);

			return filesMap;
		},
	});
