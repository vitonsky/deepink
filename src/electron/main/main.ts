import { app, Menu } from 'electron';
import { MainWindowAPI, openMainWindow } from 'src/windows/main/main';
import { isDevMode } from '@electron/utils/app';
import { openUrlWithExternalBrowser } from '@electron/utils/shell';

import { createAppMenu } from './createAppMenu';

const onShutdown = (callback: () => any) => {
	process.once('beforeExit', callback);
	process.once('SIGTERM', callback);
	process.once('SIGINT', callback);

	return () => {
		process.off('beforeExit', callback);
		process.off('SIGTERM', callback);
		process.off('SIGINT', callback);
	};
};

export class MainProcess {
	public async start() {
		// TODO: pass arguments when take a lock
		const gotTheLock = app.requestSingleInstanceLock();

		// Quit instantly if can't get lock
		if (!gotTheLock) {
			console.log('Close app because another instance is already opened');
			app.quit();
			return;
		}

		// Init app
		this.setListeners();
		Menu.setApplicationMenu(createAppMenu());
		app.whenReady().then(this.onReady);
	}

	private window: MainWindowAPI | null = null;
	private async onReady() {
		console.log('App ready');

		if (isDevMode()) {
			console.log('Install dev tools');

			const { installExtension, REACT_DEVELOPER_TOOLS, REDUX_DEVTOOLS } =
				await import(
					// eslint-disable-next-line spellcheck/spell-checker
					'electron-devtools-installer'
				);

			await Promise.all(
				[REACT_DEVELOPER_TOOLS, REDUX_DEVTOOLS].map((extension) =>
					installExtension(extension, {
						loadExtensionOptions: { allowFileAccess: true },
					})
						.then((ext) => console.log(`Added Extension:  ${ext.name}`))
						.catch((err) => console.log('An error occurred: ', err)),
				),
			);
		}

		const windowControls = await openMainWindow();
		this.window = windowControls;

		// Force app shutdown
		onShutdown(() => {
			windowControls.quit();
			app.exit();
			this.window = null;
		});
	}

	private setListeners() {
		// Handle attempts to open new instance of app
		app.on(
			'second-instance',
			(_event, _commandLine, _workingDirectory, additionalData) => {
				// Print out data received from the second instance.
				console.log(
					'Another app instance just requested a single instance lock',
					additionalData,
				);

				// Someone tried to run a second instance, we should focus our window.
				if (this.window) {
					this.window.openWindow();
				}
			},
		);

		// Open external links in default browser
		app.on('web-contents-created', (_event, webContents) => {
			webContents.setWindowOpenHandler(({ url }) => {
				console.log(
					'Prevent open new window, and open URL as external link',
					url,
				);

				openUrlWithExternalBrowser(url);

				return { action: 'deny' };
			});

			webContents.on('will-navigate', (event, url) => {
				if (url !== webContents.getURL()) {
					console.log(
						'Prevent navigation in main window, and open URL as external link',
						url,
					);

					event.preventDefault();
					openUrlWithExternalBrowser(url);
				}
			});
		});
	}
}
