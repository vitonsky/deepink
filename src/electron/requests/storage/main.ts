import { IpcMainInvokeEvent } from 'electron';
import { existsSync, statSync } from 'fs';
import path from 'path';
import recursive from 'recursive-readdir';
import { ApiToHandlers } from '@electron/utils/ipc';
import { recoveryAtomicFile, writeFileAtomic } from '@utils/files';

import { getUserDataPath, joinPath } from '../../utils/files';
import { ipcMainHandler } from '../../utils/ipc/ipcMainHandler';

import { mkdir, readFile, rm } from 'fs/promises';
import { storageChannel, StorageChannelAPI } from '.';

export const createStorageBackend = () => {
	const getScopedPath = (subdir: string | undefined, path?: string) =>
		path ? joinPath(getUserDataPath(subdir), path) : getUserDataPath(subdir);

	return {
		async upload({ req: [id, buffer, subdir] }) {
			const filePath = getScopedPath(subdir, id);
			await mkdir(path.dirname(filePath), { recursive: true });
			await writeFileAtomic(filePath, Buffer.from(new Uint8Array(buffer)));
		},

		async get({ req: [id, subdir] }) {
			const filePath = getScopedPath(subdir, id);

			recoveryAtomicFile(filePath);

			if (!existsSync(filePath) || !statSync(filePath).isFile()) return null;

			const buffer = await readFile(filePath);
			return new Uint8Array(buffer).buffer;
		},

		async delete({ req: [ids, subdir] }) {
			for (const id of ids) {
				const filePath = getScopedPath(subdir, id);

				if (!existsSync(filePath)) {
					console.debug('Not found file', filePath);
					continue;
				}

				await rm(filePath, { force: true, recursive: true });
				console.debug('Removed file', filePath);
			}
		},

		async list({ req: [subdir] }) {
			const filesDir = getScopedPath(subdir);

			if (!existsSync(filesDir)) return [];

			const files = await recursive(filesDir);
			return files.map((path) =>
				// Remove root path
				path.slice(filesDir.length),
			);
		},
	} satisfies ApiToHandlers<StorageChannelAPI, IpcMainInvokeEvent>;
};

export const enableStorage = () =>
	storageChannel.server(ipcMainHandler, createStorageBackend());
