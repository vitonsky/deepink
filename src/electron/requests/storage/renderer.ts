import { IEncryptionController } from '@core/encryption';
import { IFilesStorage } from '@core/features/files';

import { ipcRendererFetcher } from '../../utils/ipc/ipcRendererFetcher';

import { storageChannel, StorageChannelAPI } from '.';

export const storageApi = storageChannel.client(ipcRendererFetcher);

// TODO: transparently resolve all file names like it is absolute paths
export class ElectronFilesController implements IFilesStorage {
	constructor(
		private readonly storageApi: StorageChannelAPI,
		private readonly subdirectory: string,
		private readonly encryption?: IEncryptionController,
	) {
		this.subdirectory = subdirectory;
		this.encryption = encryption;
	}

	public async write(id: string, buffer: ArrayBuffer) {
		const encryptedBuffer = this.encryption
			? await this.encryption.encrypt(buffer)
			: buffer;
		return this.storageApi.upload(id, encryptedBuffer, this.subdirectory);
	}

	public async get(id: string) {
		return this.storageApi.get(id, this.subdirectory).then((buffer) => {
			// Don't handle empty data
			if (!buffer) return buffer;

			if (!this.encryption) return buffer;
			return this.encryption.decrypt(buffer);
		});
	}

	public async delete(ids: string[]) {
		return this.storageApi.delete(ids, this.subdirectory);
	}

	public async list() {
		return this.storageApi.list(this.subdirectory);
	}
}
