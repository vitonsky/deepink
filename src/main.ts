import { app, Menu } from 'electron';
import { openMainWindow } from 'src/windows/main/main';
import { isDevMode } from '@electron/utils/app';
import { getResourcesPath } from '@electron/utils/files';
import { isPlatform } from '@electron/utils/platform';

import { createAppMenu } from './createAppMenu';

console.log({
	isDev: isDevMode(),
	appDir: app.getAppPath(),
	resourcesPath: getResourcesPath(),
});

Menu.setApplicationMenu(null);
app.whenReady().then(async () => {
	console.log('App ready');

	if (isDevMode()) {
		console.log('Install dev tools');

		const { installExtension, REACT_DEVELOPER_TOOLS, REDUX_DEVTOOLS } = await import(
			// eslint-disable-next-line spellcheck/spell-checker
			'electron-devtools-installer'
		);

		await Promise.all(
			[REACT_DEVELOPER_TOOLS, REDUX_DEVTOOLS].map((extension) =>
				installExtension(extension, {
					loadExtensionOptions: { allowFileAccess: true },
				})
					.then((ext) => console.log(`Added Extension:  ${ext.name}`))
					.catch((err) => console.log('An error occurred: ', err)),
			),
		);
	}

	openMainWindow();
});

app.on('ready', () => {
	// Add application menu for mac
	if (isPlatform('darwin')) {
		Menu.setApplicationMenu(createAppMenu());
	}
});
