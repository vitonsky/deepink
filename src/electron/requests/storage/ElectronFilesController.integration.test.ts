import { vol } from 'memfs';

import { enableStorage } from './main';
import { ElectronFilesController, storageApi } from './renderer';

vi.mock('fs', () => vi.importActual('@mocks/fs'));
vi.mock('fs/promises', () => vi.importActual('@mocks/fs/promises'));
vi.mock('recursive-readdir', () => vi.importActual('@mocks/recursive-readdir'));
vi.mock('electron', () => vi.importActual('@mocks/electron'));

const storageHandlerCleanup = enableStorage();
const filesController = new ElectronFilesController(storageApi, 'profileDir');

const getBufferFromText = (text: string) => new Uint8Array(Buffer.from(text)).buffer;

test('Write files', async () => {
	expect(vol.readdirSync('/', { recursive: true })).toEqual([]);

	await filesController.write('test.txt', Buffer.from('File text content'));
	await filesController.write(
		'/foo/bar/baz/test.txt',
		Buffer.from('File text content'),
	);

	expect(vol.readdirSync('/', { recursive: true })).toMatchSnapshot(
		'FS structure after writing',
	);
	await expect(filesController.list()).resolves.toMatchSnapshot(
		'Controller files list after writing',
	);
});

test('Check files content', async () => {
	await expect(
		filesController.list().then((paths) =>
			Promise.all(
				paths.map((path) =>
					filesController.get(path).then((buffer) => {
						if (!buffer) throw new Error('Buffer not found');
						return {
							path,
							content: Buffer.from(buffer).toString(),
						};
					}),
				),
			),
		),
	).resolves.toMatchSnapshot('Files content');
});

test('File may be accessed via absolute and relevant paths both', async () => {
	await expect(filesController.get('test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
	await expect(filesController.get('/test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
	await expect(filesController.get('////test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
	await expect(filesController.get('/foo/../test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);

	await expect(filesController.get('/foo/bar/baz/test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
	await expect(filesController.get('./foo/bar/baz/test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
	await expect(filesController.get('foo/bar/baz/test.txt')).resolves.toStrictEqual(
		getBufferFromText('File text content'),
	);
});

describe('Deletion tests', () => {
	const files = new ElectronFilesController(storageApi, 'deletionTests');

	test('Add files', async () => {
		await expect(files.write('/root.bytes', new Uint8Array())).resolves.not.toThrow();
		await expect(files.write('/bytes1', new Uint8Array())).resolves.not.toThrow();
		await expect(files.write('/bytes2', new Uint8Array())).resolves.not.toThrow();
		await expect(files.write('/bytes3', new Uint8Array())).resolves.not.toThrow();
		await expect(files.write('/foo/bytes1', new Uint8Array())).resolves.not.toThrow();
		await expect(files.write('/foo/bytes2', new Uint8Array())).resolves.not.toThrow();
		await expect(
			files.write('/foo/bar/bytes1', new Uint8Array()),
		).resolves.not.toThrow();
		await expect(
			files.write('/foo/bar/bytes2', new Uint8Array()),
		).resolves.not.toThrow();
	});

	test('Single file may be deleted', async () => {
		await expect(files.list()).resolves.toContain('/bytes1');
		await files.delete(['bytes1']);
		await expect(files.list()).resolves.not.toContain('/bytes1');
	});

	test('Few files may be deleted', async () => {
		await expect(files.list()).resolves.toContain('/bytes2');
		await expect(files.list()).resolves.toContain('/bytes3');

		await files.delete(['bytes2', '/bytes3']);

		await expect(files.list()).resolves.not.toContain('/bytes2');
		await expect(files.list()).resolves.not.toContain('/bytes3');
	});

	test('Directory deletion deletes all its content', async () => {
		await expect(files.list()).resolves.toContain('/foo/bar/bytes1');
		await expect(files.list()).resolves.toContain('/foo/bar/bytes2');

		await files.delete(['/foo/bar']);
		await expect(files.list()).resolves.not.toContain('/foo/bar/bytes1');
		await expect(files.list()).resolves.not.toContain('/foo/bar/bytes2');
	});

	test('Root directory may be deleted', async () => {
		await expect(files.list()).resolves.not.toEqual([]);

		await files.delete(['/']);
		await expect(files.list()).resolves.toEqual([]);
	});
});

describe('Many clients', () => {
	const files1 = new ElectronFilesController(storageApi, 'profile1');
	const files2 = new ElectronFilesController(storageApi, 'profile2');

	test('Available files is scoped by client root', async () => {
		await files1.write('test.bytes', new Uint8Array(1));

		await expect(files1.list()).resolves.toEqual(['/test.bytes']);
		await expect(files2.list()).resolves.toEqual([]);

		await files2.write('test.bytes', new Uint8Array(2));

		await expect(files1.get('test.bytes')).resolves.toStrictEqual(
			new Uint8Array(1).buffer,
		);
		await expect(files2.get('test.bytes')).resolves.toStrictEqual(
			new Uint8Array(2).buffer,
		);
	});

	test("Client can't traverse out of it's root", async () => {
		await expect(files1.get('../profile2/test.bytes')).rejects.toThrowError(
			'Resolved path is out of root directory',
		);
	});
});

test('Path traversal out of scope directory must not be possible', async () => {
	await expect(filesController.get('/')).resolves.toBeNull();
	await expect(filesController.write('/', new ArrayBuffer(1))).rejects.toThrowError(
		"EISDIR: illegal operation on a directory, open '/home/userData/appDir/profileDir'",
	);

	// Just to ensure we can read available files
	vol.writeFileSync('/home/userData/appDir/profileDir/secret.txt', 'Secret data');
	await expect(filesController.get('./secret.txt')).resolves.toBeInstanceOf(
		ArrayBuffer,
	);

	// Read file out of scoped directory
	// Scope is defined by a client
	vol.writeFileSync('/home/userData/appDir/secret.txt', 'Secret data');
	await expect(filesController.get('../secret.txt')).rejects.toThrowError(
		'Resolved path is out of root directory',
	);

	// All ops respects the limitation
	await expect(filesController.delete(['../secret.txt'])).rejects.toThrowError(
		'Resolved path is out of root directory',
	);
	await expect(
		filesController.write('../secret.txt', new Uint8Array()),
	).rejects.toThrowError('Resolved path is out of root directory');

	// Read file out of root directory
	vol.writeFileSync('/home/userData/secret.txt', 'Secret data');
	await expect(filesController.get('../../secret.txt')).rejects.toThrowError(
		'Resolved path is out of root directory',
	);
});

test('When storage is disposed, any ops in controller throws error', async () => {
	// Clean handler
	storageHandlerCleanup();
	await expect(filesController.list()).rejects.toThrow(
		'No handler registered for storage.list',
	);
	await expect(filesController.get('/')).rejects.toThrow(
		'No handler registered for storage.get',
	);
	await expect(filesController.write('/', new ArrayBuffer(1))).rejects.toThrow(
		'No handler registered for storage.upload',
	);
	await expect(filesController.delete(['/'])).rejects.toThrow(
		'No handler registered for storage.delete',
	);

	// Setup handler
	const storageHandlerCleanup2 = enableStorage();
	await expect(filesController.list()).resolves.toContainEqual('/test.txt');

	// Clear handler again
	storageHandlerCleanup2();
	await expect(filesController.list()).rejects.toThrow(
		'No handler registered for storage.list',
	);
});
