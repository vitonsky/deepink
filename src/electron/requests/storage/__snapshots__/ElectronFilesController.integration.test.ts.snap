// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check files content > Files content 1`] = `
[
  {
    "content": "File text content",
    "path": "/test.txt",
  },
  {
    "content": "File text content",
    "path": "/foo/bar/baz/test.txt",
  },
]
`;

exports[`Write files > Controller files list after writing 1`] = `
[
  "/test.txt",
  "/foo/bar/baz/test.txt",
]
`;

exports[`Write files > FS structure after writing 1`] = `
[
  "/home/userData/appDir",
  "/home/userData/appDir/profileDir/foo/bar",
  "/home/userData/appDir/profileDir/foo/bar/baz",
  "/home/userData/appDir/profileDir/foo",
  "home",
  "/home/userData/appDir/profileDir",
  "/home/userData/appDir/profileDir/test.txt",
  "/home/userData/appDir/profileDir/foo/bar/baz/test.txt",
  "/home/userData",
]
`;
